src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -257,7 +259,7 @@ public bool Equals([NotNullWhen(true)] string? text, StringComparison comparison
         {
             if (text == null)
             {
-                return false;
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
```

Comment:
If we are going to throw NullArg here, the `text` parameter needs to be updated to say `string text`, so callers know they shouldn't pass `null` in here.

Reply:
`text` cannot be non-nullable in `Equals(string? text)` (by interface contract): https://github.com/dotnet/runtime/pull/57395#discussion_r688975416

So it might be strange to make it non-nullable for `Equals([NotNullWhen(true)] string? text, StringComparison comparisonType)`: https://github.com/dotnet/runtime/pull/57395#discussion_r688974246

This was discussed in the comments above.

I can make `text` non-nullable, but only in `Equals([NotNullWhen(true)] string? text, StringComparison comparisonType)`

Reply:
I don't understand why we'd throw here.  It's equals or it's not... what necessitate making it yes,  no,  or throw?

Reply:
I think, returning `false` from `new StringSegment(null).Equals((string?)null);` would be even less obvious.

Reply:
> No idea why it doesn't return true though.

Didn't you just change it to return false a few days ago and before that it was throwing?

We should make it return whatever value makes sense, given the combination of the current instance and the provided value.  Throwing doesn't make sense,  at least not to me, given that Equals in general allows null arguments.

Reply:
StringSegment has an implicit conversion from string. What about logic equivalent to:

```cs
public bool Equals([NotNullWhen(true)] string? text, StringComparison comparisonType)
{
    if (text == null)
    {
        StringSegment other = text;
        return Equals(other);
    }

    /// ...
}
```

You should also update the implicit conversion to allow a nullable string (~line 316):

```diff
- public static implicit operator StringSegment(string value) => new StringSegment(value);
+ public static implicit operator StringSegment(string? value) => new StringSegment(value);
```

Reply:
> What about logic equivalent to

This is what I meant by "Presumably segment.Equals(string) should behave the same as segment.Equals(new StringSegment(string))...?"

Reply:
> Didn't you just change it to return false a few days ago and before that it was throwing?

Yep. I thought `new StringSegment(null)` was illegal. In this PR I didn't want to change behavior again without your blessing, so just reverted Equals implementation to what was originally there.

> Presumably segment.Equals(string) should behave the same as segment.Equals(new StringSegment(string))...?

Makes sense. Updated the implementation and tests so that Equals never throws.

> You should also update the implicit conversion to allow a nullable string (~line 316):

Thank you, missed that one.